src/components/vm/vmMigrateDialog.jsx

```diff
@@ -0,0 +1,206 @@
+
+const _ = cockpit.gettext;
+
+const DestUriRow = ({ validationFailed, destUri, setDestUri }) => {
+    return (
+        <FormGroup label={_("Destination URI")} fieldId="dest-uri-input"
+                   helperTextInvalid={validationFailed.destUri}
+                   validated={validationFailed.destUri ? "error" : "default"}>
+            <TextInput id='dest-uri-input'
+                       validated={validationFailed.destUri ? "error" : "default"}
+                       value={destUri}
+                       placeholder={cockpit.format(_("Example, $0"), "qemu+ssh://192.0.2.16/system")}
+                       onChange={setDestUri} />
+        </FormGroup>
+    );
+};
+
+const OptionsRow = ({ temporary, live, setTemporary, setLive, setStorage }) => {
+    return (
+        <FormGroup label={_("Options")}
+                   hasNoPaddingTop>
+            <Checkbox id="live"
+                      isChecked={live}
+                      label={_("Run on destination")}
+                      onChange={b => {
+                          setLive(b);
+                          if (!b)
+                              setStorage("nocopy");
+                      }} />
+            {live && <Checkbox id="temporary"
+                               isChecked={temporary}
+                               label={
+                                   <Flex spaceItems={{ default: 'spaceItemsSm' }} alignItems={{ default: 'alignItemsCenter' }}>
+                                       <span>{_("Move temporarily")}</span>
+                                       <Tooltip position="top"
+                                                content={
+                                                    <>
+                                                        <p>{_("By default, the migrated VM config is removed from the source host, and saved persistently on the destination host. The destination host is considered the new home of the VM.")}</p>
+                                                        <p>{_("If 'temporary' is selected, the migration is considered only a temporary move: the source host maintains a copy of the VM config, and the running copy moved to the destination is only transient, and will disappear when it is shutdown.")}</p>
+                                                    </>}>
+                                           <OutlinedQuestionCircleIcon />
+                                       </Tooltip>
+                                   </Flex>
+                               }
+                               onChange={setTemporary} />
+            }
+        </FormGroup>
+    );
+};
+
+const StorageRow = ({ storage, setStorage, live }) => {
+    const copyDisabled = !live;
+    let copyRadio = (
+        <Checkbox id="copy"
+                  label={
+                      <Flex spaceItems={{ default: 'spaceItemsSm' }} alignItems={{ default: 'alignItemsCenter' }}>
+                          <span>{_("Copy storage")}</span>
```

Comment:
When would you ever uncheck this? How is a VM supposed to be able to start without its disks?

Is this a hack because we don't know whether the VM's storage already lives on shared storage or not, and the user has to pick the right setting? Then the tooltip should point that out. If it can really work with both, please explain the "why?" for both cases in the tooltip.

I see that the option gets disabled for non-live mode, but that also seems upside down. I see that if I migrate an non-running VM, that may make some sense without storage. But migrating a live VM without storage really doesn't.

Reply:
After reading the code, it smells like this does not apply to the VM's "primary" disk, but only additionally attached storage? That would explain why it's an option at all, but not why it's dynamically coupled to "live"?

Reply:
With or without this option checked, shared storage is always _transferred_ (just definition of shared disks is transferred)
This option only specifies, if **in addition** to the shared storage, it should also copy all of the non-shared non-readonly disks. (so it transfers both the disk defitinion and the data).
That's why I'm not really happy with the Storage wording in this example:
https://matrix.org/_matrix/media/r0/download/matrix.org/eAKISkjndstziZBEeHWQdwbw/Screen%20Shot%202021-04-30%20at%2011.23.02.png
As it would lead user to believe that they either have to have all disks on shared storage, or have all disk be copied.
While in reality, they can have som disks on shared storage, and some disks locally, and they only choose if in addition to shared storage they want the local disks to be copied.
@KKoukiou What do you think about it?

Reply:
Agreed, that wording is not at all good yet -- it's both too long, and also does not point out that difference. But there seems to be some principal misunderstanding still: This morning on IRC it sounded like you need to *tell* libvirt whether your storages are shared with the target machine (nfs, cephfs, rsync, syncthing, etc.), or whether they are local; because it is pretty well impossible to detect that automatically.

But here you make a different claim: That libvirt already knows which disks are shared and which not (how?), and that this is about copying the "non-shared" disks. I think we are talking about different definitions of what a "shared" disk is?

Reply:
Ah no. I was wrong. Yes, libvirt doesn't know about it. And if user chooses to use "copy storage" option, but there is one NFS disk, it can lead to a failure. But the main situation I was trying to solve here is:
_What if VM has 2 disks: One disk is a local file disk, the other one is a shared storage disk (e.g. nfs). User wants to migrate both of them._ 

Using the first option (without --copy-storage-all) would result that the local file disk is not copied.
Using the second option (with --copy-storage-all) would result in libvirt trying to also copy the NFS disk, which would lead to an error.

But maybe I'm overthinking

Reply:
No, you are not -- that's exactly the kind of scenario that we do need to sort out. This could happen, and we can't detect it, so with the currently available data the only way seems to be to add a warning that if you mix them, you can't migrate.

Reply:
Right now, the only thing is to add a warning.
Figuring out what disks are shared and what disks are local can be complicated sometimes (like when a disk is hidden behind a mount point). And there it many ways how to set up shared storage, and it's constantly changing what libvirt considers shared storage, so even if we wrote some function to detect shared storage disks from local file disk, it would eventually divert from libvirt.

Reply:
Right, let's not get down into the rabbit hole of trying to detect shared-ness. The one thing that we must prevent is trying the migration and deleting the local instance even though the remote instance is broken (because it has no proper storage). If that always causes some error message anyway, then it's fine I think -- admins should at least raise an eyebrow if the only two options are "all storage is shared" and "all storage gets copied", and there is no "some" option.